### PR TITLE
233 analytics tool

### LINKED ISSUE #233: [FEATURE] - Analytics tool
**Issue Description**
In order to follow the number of visitors to our website, or reading the newsletters, we need a gem that will take care of it.

- [x] Install https://rubygems.org/gems/ahoy_matey
- [x] Update the admin console with required requests (visits) in order to follow it as an admin (Update Blazer requests) -> Could be done later by @alexstan57
- [x] View count on landing page
- [x] View count on each blog/newsletter
- [x] View count on each trip request

**Desired Behaviour**
See how to disregard visits done by bots

**Acceptance Criteria**
...

**Additional Considerations**
...

**Resources**
https://www.youtube.com/watch?v=G_T17DopsOk

**Impact**
...

**Related Issues**
Requires #200

### PR DESCRIPTION
**Pull request by christinasugiono**
image

closes #233

### CODE REVIEW

**Comment by christinasugiono** [comment]
Newly created article automatically included in the statistic :D

**Comment by christinasugiono** [comment]
**How To Add New Tracker**

Put this code in the controller of the page you want to track:

- With properties
`ahoy.track "Viewed Blog", blog_id: id`
- Without properties
`ahoy.track "Viewed Page"`

**Comment by christinasugiono** [comment]
## Statistics (Blazer)
### Total visitor count
```
SELECT 
    name,
    COUNT(*)
FROM ahoy_events
GROUP BY name
```

### All blogs visitor count
```
WITH blog_visit_count AS (
    SELECT (properties->>'blog_id')::integer AS blog_id,
           COUNT(*) AS view_count
    FROM ahoy_events
    WHERE name = 'Viewed blog'
    GROUP BY (properties->>'blog_id')::integer
)

SELECT b.title, bvc.view_count
FROM blogs b
LEFT JOIN blog_visit_count bvc ON b.id = bvc.blog_id;
```

### Pages visitor growth
```
SELECT 
    date_trunc({period}, time)::date AS period, 
    name,
    COUNT(*) AS visitor_count 
FROM ahoy_events 
GROUP BY period, name
ORDER BY period
```

### All blogs visitor growth
```
WITH blog_monthly_visit AS (
    SELECT 
        date_trunc({period}, time)::date AS period,
        (properties->>'blog_id')::integer AS blog_id,
        COUNT(*) AS view_count
    FROM ahoy_events
    WHERE name = 'Viewed blog'
    GROUP BY period, blog_id
    ORDER BY period
)

SELECT
    period,
    b.title,
    view_count
FROM blog_monthly_visit bmv
LEFT JOIN blogs b
ON b.id = bmv.blog_id
```

### Blog visitor growth
```
WITH blog_monthly_visit AS (
    SELECT 
        date_trunc({period}, time)::date AS period,
        (properties->>'blog_id')::integer AS blog_id,
        COUNT(*) AS view_count
    FROM ahoy_events
    WHERE name = 'Viewed blog'
    GROUP BY period, blog_id
    ORDER BY period
)

SELECT
    period,
    b.title,
    view_count
FROM blog_monthly_visit bmv
LEFT JOIN blogs b
ON b.id = bmv.blog_id
WHERE b.id = {blog}
```

**Comment by christinasugiono** [comment]
To test the `geolocation` of `Ahoy::Visit`

- Run `ngrok` [`documentation`](https://ngrok.com/docs/getting-started/)
   `localhost` won't give enough information on the IP address for geocode, so we use `ngrok`
- Run `rake solid_queue:start`
- Visit landing page/trip request/blogs

File: app/models/ahoy/event.rb
Comment (alexstan67) [inline app/models/ahoy/event.rb:1]: Is this class necessary?

File: app/models/ahoy/visit.rb
Comment (alexstan67) [inline app/models/ahoy/visit.rb:1]: Is this class necessary?

File: app/views/blogs/_form.html.erb
Comment (alexstan67) [inline app/views/blogs/_form.html.erb:1]: Why did you change this one? For example, if you edit a drafted blog, the check boxes below disappeared... Not working anymore...

File: config/locales/default_en.yml [hunk 1]
```diff
@@ -48,4 +48,5 @@ en:
     privacy_policy:
     copyright: "All rights reserved"
     console: Console
+    statistic: Statistics
```
Comment (alexstan67) [inline config/locales/default_en.yml:51]:
```suggestion
```

File: config/initializers/ahoy.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+class Ahoy::Store < Ahoy::DatabaseStore
+  def authenticate(data)
```
Comment (alexstan67) [inline config/initializers/ahoy.rb:2]:
```suggestion
```

File: config/initializers/ahoy.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    # disables automatic linking of visits and users
```
Comment (alexstan67) [inline config/initializers/ahoy.rb:3]:
```suggestion
```

File: config/initializers/ahoy.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  end
```
Comment (alexstan67) [inline config/initializers/ahoy.rb:4]:
```suggestion
```

File: config/locales/default_fr.yml [hunk 5]
```diff
@@ -96,4 +96,5 @@ fr:
     privacy_policy: "Politique de confidentialité"
     copyright: "Tous droits réservés"
     console: "Console"
+    statistic: "Statistiques"
```
Comment (alexstan67) [inline config/locales/default_fr.yml:99]:
```suggestion
```

File: db/migrate/20240508075747_create_ahoy_visits_and_events.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  def change
+    create_table :ahoy_visits do |t|
+      t.string :visit_token
+      t.string :visitor_token
+
+      # the rest are recommended but optional
+      # simply remove any you don't want
+
+      # user
+      t.references :user
```
Comment (alexstan67) [inline db/migrate/20240508075747_create_ahoy_visits_and_events.rb:11]:
```suggestion
```

File: db/migrate/20240508075747_create_ahoy_visits_and_events.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+      # utm parameters
+      t.string :utm_source
+      t.string :utm_medium
+      t.string :utm_term
+      t.string :utm_content
+      t.string :utm_campaign
+
+      # native apps
+      t.string :app_version
```
Comment (alexstan67) [inline db/migrate/20240508075747_create_ahoy_visits_and_events.rb:40]:
```suggestion
      #t.string :app_version
```

File: db/migrate/20240508075747_create_ahoy_visits_and_events.rb [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+      t.string :os_version
```
Comment (alexstan67) [inline db/migrate/20240508075747_create_ahoy_visits_and_events.rb:41]:
```suggestion
      #t.string :os_version
```

File: db/migrate/20240508075747_create_ahoy_visits_and_events.rb [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+      t.string :platform
```
Comment (alexstan67) [inline db/migrate/20240508075747_create_ahoy_visits_and_events.rb:42]:
```suggestion
      #t.string :platform
```

File: db/migrate/20240508075747_create_ahoy_visits_and_events.rb [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+      t.datetime :started_at
+    end
+
+    add_index :ahoy_visits, :visit_token, unique: true
+    add_index :ahoy_visits, [:visitor_token, :started_at]
+
+    create_table :ahoy_events do |t|
+      t.references :visit
+      t.references :user
```
Comment (alexstan67) [inline db/migrate/20240508075747_create_ahoy_visits_and_events.rb:52]:
```suggestion
      #t.references :user
```

File: test/system/users_test.rb
Comment (alexstan67) [inline test/system/users_test.rb:1]: Ok, but in this case should we replace it with a controller users test?

File: app/models/user.rb [hunk 11]
```diff
@@ -9,6 +9,7 @@ class User < ApplicationRecord
   has_many :followings, dependent: :destroy, foreign_key: :follower_id,
                         class_name: 'Follower', inverse_of: :follower
   has_many :audit_logs, dependent: :destroy
+  has_many :visits, class_name: 'Ahoy::Visit'
```
Comment (alexstan67) [inline app/models/user.rb:12]:
```suggestion
```

**Review by alexstan67 — CHANGES_REQUESTED** [review]
I have only 3 general comments:
- We want to activate the local geocoding of countries (not cities): https://www.rubydoc.info/gems/ahoy_matey/5.1.0#local-geocoding
  - it implies to update the `Readme.md`
- We don't want to track users, so we remove the links between visits, events and users (GDPR)
- I'm not sure why you made modification on Blogs?
Happy to discuss, ++

File: app/models/ahoy/event.rb
Comment (christinasugiono) [reply app/models/ahoy/event.rb:1]: I assume yes, because its generated during the gem ahoy setup rails generate ahoy:install :D

**Review by alexstan67 — APPROVED** [review]

File: test/system/users_test.rb
Comment (alexstan67) [inline test/system/users_test.rb:1]: What do we do with this one?

File: app/views/blogs/_form.html.erb
Comment (alexstan67) [inline app/views/blogs/_form.html.erb:1]: Is there a link with analytics tools? Is it commit from the fix branch?

**Review by alexstan67 — COMMENTED** [review]
Let's merge!

**Pull request merged**